Fdtgrep and libfdt_extra

sjg20 opened the pull request:
This is the long-awaited fdtgrep tool with the libfdt additions to make it work.

These changes were rejected from the main libfdt tree, so are housed here.

sjg20 committed:
Set up build files

Bring in some build files from the dtc repo, so we can use the same
approach.

Changes made:
- Drop Python-related stuff
- Use FDT_VERSION/fdt_version instead of DTC_VERSION/dtc_version
- Drop unwanted binary and test-file rules
- Add license headers
- Fix some of the easier checkpatch warnings

For license headers, SPDX is used, with licences matching what is there.
Where there is no licence, GPL-2.0+ is used.

With this, both 'make' and 'make check' work, although they do nothing
useful for now.

The dtc[1] commit used is:

   ccf1f62 ("libfdt: Fix a typo in libfdt.h")

[1] git://git.kernel.org/pub/scm/utils/dtc/dtc.git

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
Add license information

Add a README for licensing, along with the two licenses used in this
project.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
Add functions to find regions in an FDT

Given a set of nodes and properties, find the regions of the device tree
which describe those parts.

Put this feature in a new libfdt_extra library.

A test is provided which builds a tree while tracking where the regions
should be, then calls fdt_first/next_region() to make sure that it agrees.

Further tests will come as part of fdtgrep.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
Add fdtgrep to grep and subset FDTs

fdtgrep operates like a grep for device trees, working on binary .dtb
files.

fdtgrep can find nodes by name or compatible string. It can find properties
by name. Using a list of nodes/properties/compatible strings to match,
fdtgrep creates either .dts text output containing just those matches, or
.dtb output. In the .dtb case, the output is a valid FDT which only
includes the selected nodes/properties. This is useful in
resource-constrained systems where a full FDT is too large to fit in
available memory (e.g. U-Boot SPL).

fdtgrep also supports producing raw binary output, without the normal FDT
headers. This can be useful for hashing part of an FDT and making sure
that that part does not change, even if other parts do. This makes it
possible to detect an important change, while ignoring a change that is
of no interest.

At its simplest fdtgrep can be used as follows;

   fdtgrep /holiday tests/grep.dtb

and it produces just that node (with a skeleton to hold it):

/ {
    holiday {
        compatible = "ixtapa", "mexico";
        weather = "sunny";
        status = "okay";
    };
};

The binary form of this is:

   fdtgrep -n /holiday -O dtb tests/grep.dtb

which produces a valid DTB file with just the above nodes.

Various options are provided to search for properties, to ensure that
nodes/properties are not present, and to search for nodes by compatible
string.

Signed-off-by: Simon Glass <sjg@chromium.org>

robherring commented on License/README:
```diff
@@ -0,0 +1,51 @@
+
+	SPDX-License-Identifier: GPL-2.0+ OR BSD-3-Clause
+
+you can choose between GPL-2.0+ and BSD-3-Clause licensing.
```
Here you say 3 clause, but the license file is 2 clause.

sjg20 replied:
fixed

robherring commented on Makefile:
dtc is trying to move to meson. Can't we do that from the start here? Then we're not trying to maintain compatibility when meson support is added.

sjg20 replied:
Hmm I was wondering about that...I will take a look

sjg20 replied:
done

sjg20 committed:
Add meson files

Provide meson files for building.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 commented:
I notice that meson tests don't work in the dtc repo

sjg20 commented:
Just checking if there is anything else to do here?

robherring commented:
My suggestion was not to have both meson and Makefiles, but just meson. Maintaining both is a pain depending to what extent you care about both supporting all the same features/setups. But it's your repo. If you are good with it, merge it.

sjg20 commented:
That makes more sense...but did you see my question? How do I run the tests using meson? i.e. the equivalent of 'make check'

robherring commented:
This works for me:
```
meson setup build
cd build
meson compile
meson test -v

```
It doesn't look like the result counting works right though. It seems to be a single pass or fail.

sjg20 committed:
Drop Makefiles

We are using meson now and the makefiles are in the history for reference.
Delete them.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 commented:
OK, updated, thanks for helping with the tests

robherring commented:
LGTM. Maybe merge it into 'main' instead of master?

sjg20 commented:
Hmm not sure how to change that, but I should be able to do both

robherring commented:
There's a GH UI thing to rename master. It will redirect master to main on the web, but can't migrate users that have cloned it. So best to do that before anyone clones it (besides you).

sjg20 commented:
OK thanks, I think I figured that out
